src/core/ext/transport/binder/java/io/grpc/binder/cpp/SyncServiceConnection.java

```diff
@@ -0,0 +1,71 @@
+// Copyright 2021 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package io.grpc.binder.cpp;
+
+import android.content.ComponentName;
+import android.content.Context;
+import android.content.Intent;
+import android.content.ServiceConnection;
+import android.os.IBinder;
+import android.util.Log;
+
+/* Connects to a service synchronously */
+public class SyncServiceConnection implements ServiceConnection {
+  private final String logTag = "SyncServiceConnection";
+
+  private Context mContext;
+  private IBinder mService;
+
+  public SyncServiceConnection(Context context) {
+    mContext = context;
+  }
+
+  @Override
+  public void onServiceConnected(ComponentName className, IBinder service) {
+    Log.e(logTag, "Service has connected: ");
+    synchronized (this) {
+      mService = service;
+    }
+  }
+
+  @Override
+  public void onServiceDisconnected(ComponentName className) {
+    Log.e(logTag, "Service has disconnected: ");
+  }
+
+  public void tryConnect() {
+    synchronized (this) {
+      Intent intent = new Intent("grpc.io.action.BIND");
+      // TODO(mingcl): The component name is currently hard-coded here and should be changed
+      // manually before compile. We should pump the component name from C++ to here instead after
+      // we have a server ready for integration test.
+      ComponentName compName = new ComponentName("redacted", "redacted");
+      intent.setComponent(compName);
+      // Will return true if the system is in the process of bringing up a service that your client
+      // has permission to bind to; false if the system couldn't find the service or if your client
+      // doesn't have permission to bind to it
+      boolean result = mContext.bindService(intent, this, Context.BIND_AUTO_CREATE);
+      if (result) {
+        Log.e(logTag, "bindService ok");
+      } else {
+        Log.e(logTag, "bindService not ok");
+      }
+    }
+  }
+
+  public IBinder getIBinder() {
```

Comment:
Should we return Optional here for the case service is disconnected?

Reply:
Added a TODO in NativeConnectionHelper